### PR TITLE
Initial documentation pass on API module

### PR DESCRIPTION
**Pull request by dxslly**
- Added README to API module
- Added a couple overloaded methods to CredentialClient for common cases.
- Added examples to javadoc on CredentialClient.  
- Removed unused callback interfaces
- Renamed HintRetrieveRequest.of() to .fromAuthMethods() to match Credential request.

### CODE REVIEW

File: api/README.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+Receive the results and check if a hint was returned:
+1. If a hint was returned, it is recommended to check if an account with a matching identifier
+   already exists and intelligently switch to a log-in flow. If no existing account matches there
+   are several branches of actions your app may want to take:
+    1. If enough information is present, you may automatically create a new account for the user.
+       The amount of information required will depend on the authentication method and personal
+       requirements. Two common cases when you may want to do this if an Id Token was returned or a
+       generated password.
+    1. If more information is needed to create an account, it is recommended to use the given
+       information to skip manual entry and have the use finish account creation manually.
```
Comment (jstimes) [inline api/README.md]: have the *user* finish

File: api/README.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+1. If a hint was not returned, it is recommended fallback to manual account creation unless the
```
Comment (jstimes) [inline api/README.md]: insert "to" before "fallback"

File: api/README.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+startActivityForResult(deleteCredentialIntent, RC_DELETE_CREDENTIAL);
+```
+
+
+### Signing Out
+
+There is a common pitfall when implementing account sign-in/sign-out where an application will
+automatically sign-in a user after they have just signed out when the user's intention was likely to
+either remain signed out or switch to a new account. To resolve this, some state needs to be added
+tracking the user's previous intent and prevent the user from being automatically signed in again.
```
Comment (jstimes) [inline api/README.md]: insert "for" before tracking

File: api/java/org/openyolo/api/CredentialClient.java [hunk 4]
```diff
@@ -436,15 +607,28 @@ public CredentialDeleteResult getDeleteResult(Intent resultData) {
...
-    }
-
-    @NonNull
-    private static HintRetrieveResult createDefaultHintRetrieveResult() {
-        return HintRetrieveResult.UNKNOWN;
+    /**
+     * Disables automatically signing a user until a successful response is given to
+     * {@link CredentialClient#getCredentialRetrieveResult(Intent)}. Instead a user is required to
+     * interact with the providers (e.g. click the credential they would like to sign in) before a
+     * the credential is returned.
```
Comment (jstimes) [inline api/java/org/openyolo/api/CredentialClient.java]: remove "a" or "the"

File: api/java/org/openyolo/api/CredentialClient.java [hunk 5]
```diff
@@ -436,15 +607,28 @@ public CredentialDeleteResult getDeleteResult(Intent resultData) {
...
-    private static CredentialRetrieveResult createDefaultCredentialRetrieveResult() {
-        return new CredentialRetrieveResult.Builder(CredentialRetrieveResult.CODE_UNKNOWN)
-                .build();
-    }
-
-    @NonNull
-    private static HintRetrieveResult createDefaultHintRetrieveResult() {
-        return HintRetrieveResult.UNKNOWN;
+    /**
+     * Disables automatically signing a user until a successful response is given to
```
Comment (jstimes) [inline api/java/org/openyolo/api/CredentialClient.java]: signing *in* a user

File: api/README.md [hunk 6]
```diff
@@ -0,0 +1,208 @@
+# OpenYOLO for Android - Programmatic Credential API
+
+This is an open-source library for the OpenYOLO for Android protocol. It allows you to easily
+manipulate credentials for you application by interacting with the credential provider of the user's
```
[the remark this answers is not in the document]
Comment (dxslly) [reply api/README.md]: Done.

File: api/README.md [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+## Terminology
+
+ - [Authentication Method](https://spec.openyolo.org/openyolo-android-spec.html#authentication-methods):
+An authentication method is a mechanism by which a credential can be verified. It is represented as
+a URI of the form ```scheme://authority```. For example OpenYOLO defines several common types such
+as ``openyolo://email`` and ```openyolo://phone```.  When requesting a credential or hint your
+application will declare which authentication methods it supports.
+ - [Authentication Domain](https://spec.openyolo.org/openyolo-android-spec.html#authentication-domains):
+An authentication domain is the scope within a credential is considered to be usable. A domain
```
[the remark this answers is not in the document]
Comment (dxslly) [reply api/README.md]: Done.

File: api/README.md [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+spans a single Android application or web address. If your authentication system does not span
+multiple domains you can safely ignore this concept as a client. However for example, if your
```
[the remark this answers is not in the document]
Comment (dxslly) [reply api/README.md]: Done.

File: api/README.md [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+application offers both an app and web app you can prove this relationship to credential providers
+to allow credentials to be shared between the two. ```TOOD(dxslly): Add guide for linking
+authentication domains.```
+
+## Usage
+
+The entry point is the ```CredentialClient``` class. It is a light weight client that can be created
+as needed and offers methods to perform all standard flows. The majority of flows follow the pattern
+of sending requests via launching Activity based intents and optionally handling their results via
+the ```onActivitResult()``` call back.
```
[the remark this answers is not in the document]
Comment (dxslly) [reply api/README.md]: Done.

File: api/README.md [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+
+A typical sequence of flows for sign-in might looks as follows:
+1. Auto Sign-In: Attempt to auto sign in the user, by retrieving a saved credential.
+1. Account Creation: If no credential was found and the user is interested in signing in, assist in
+account creation by retrieving a hint.
+1. Save valid credential: If the user was successfully signed in, save the credential for future
+retrieve calls.
+1. Delete stale credential: If a retrieved credential is no longer valid, delete the credential so
+it will no be returned in future retrieve calls.
```
[the remark this answers is not in the document]
Comment (dxslly) [reply api/README.md]: Done.

File: api/README.md [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+
+### Auto Sign-In
+
+It is recommended to always try to automatically Sign-In the user in first.  This is done by
```
[the remark this answers is not in the document]
Comment (dxslly) [reply api/README.md]: Done.

File: api/README.md [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+           // user's intent such as CredentialRetrieveResult.CODE_USER_CANCELED.
+           result.getResultCode();
+       }
+    }
+}
+```
+
+### Account Creation
+
+If the user could not be automatically signed-in assisting with account creation is the next
```
[the remark this answers is not in the document]
Comment (dxslly) [reply api/README.md]: Done.

File: api/README.md [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+startActivityForResult(retrieveCredentialIntent, RC_RETRIEVE_HINT);
+```
+
+Receive the results and check if a hint was returned:
+1. If a hint was returned, it is recommended to check if an account with a matching identifier
+   already exists and intelligently switch to a log-in flow. If no existing account matches there
+   are several branches of actions your app may want to take:
+    1. If enough information is present, you may automatically create a new account for the user.
+       The amount of information required will depend on the authentication method and personal
+       requirements. Two common cases when you may want to do this if an Id Token was returned or a
```
[the remark this answers is not in the document]
Comment (dxslly) [reply api/README.md]: Removed.

File: api/README.md (hunk 1 above)
Comment (dxslly) [reply api/README.md]: Done.

File: api/README.md (hunk 2 above)
Comment (dxslly) [reply api/README.md]: Done.

File: api/README.md (hunk 3 above)
Comment (dxslly) [reply api/README.md]: Done.

File: api/java/org/openyolo/api/CredentialClient.java (hunk 5 above)
Comment (dxslly) [reply api/java/org/openyolo/api/CredentialClient.java]: Done.

File: api/java/org/openyolo/api/CredentialClient.java (hunk 4 above)
Comment (dxslly) [reply api/java/org/openyolo/api/CredentialClient.java]: Done.

File: api/README.md [hunk 14]
```diff
@@ -0,0 +1,208 @@
+# OpenYOLO for Android - Programmatic Credential API
+
+This is an open-source library for the OpenYOLO for Android protocol. It allows you to easily
```
[the remark this answers is not in the document]
Comment (dxslly) [reply api/README.md]: Done.

File: api/README.md [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+1. Prompt users after sign-in/sign up to save their credentials for future automatic sign in.
+
+## Installation
+As a pre-requisite, you will need to add the OpenYOLO Api as a dependency. For example, if your
+project uses Gradle:
+
+```groovy
+dependencies {
+    // ...
+    compile 'com.openyolo:api:xxx'
```
[the remark this answers is not in the document]
Comment (dxslly) [reply api/README.md]: Done.

File: api/README.md [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+}
+```
+
+## Terminology
+
+ - [Authentication Method](https://spec.openyolo.org/openyolo-android-spec.html#authentication-methods):
```
[the remark this answers is not in the document]
Comment (dxslly) [reply api/README.md]: Done. The spec references in the API and Protocol modules should also be updated.

**Pull request merged**